com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs

```diff
@@ -17,37 +17,117 @@ public string name
             get => m_Name;
             set => m_Name = value;
         }
+        public string categoryGuid => this.objectId;
 
         [SerializeField]
-        Guid m_CategoryGuid;
+        List<JsonRef<ShaderInput>> m_ChildObjectList = new List<JsonRef<ShaderInput>>();
 
-        public Guid categoryGuid
+        public RefValueEnumerable<ShaderInput> Children => m_ChildObjectList.SelectValue();
+
+        // We expose the object list as a HashSet of their objectIDs for faster existence checks
+        HashSet<string> m_ChildObjectIDSet = new HashSet<string>();
+
+        public int childCount => m_ChildObjectIDSet.Count;
+
+        public void InsertItemIntoCategory(ShaderInput itemToAdd, int insertionIndex = -1)
         {
-            get => m_CategoryGuid;
-            set => m_CategoryGuid = value;
+            if (itemToAdd == null)
+            {
+                AssertHelpers.Fail("Tried to insert item into category that was null.");
+                return;
+            }
+
+            if (insertionIndex == -1)
+            {
+                m_ChildObjectList.Add(itemToAdd);
+                m_ChildObjectIDSet.Add(itemToAdd.objectId);
+            }
+            else
+            {
+                m_ChildObjectList.Insert(insertionIndex, itemToAdd);
+                // Recreate the hash-set as the data structure does not allow for moving/inserting elements
+                m_ChildObjectIDSet.Clear();
+                foreach (var childObject in m_ChildObjectList)
+                {
+                    m_ChildObjectIDSet.Add(childObject.value.objectId);
+                }
+            }
         }
 
-        // We store Guids as a list out to the graph asset
-        [SerializeField]
-        List<Guid> m_ChildItemIDList;
+        public void RemoveItemFromCategory(ShaderInput itemToRemove)
+        {
+            m_ChildObjectList.Remove(itemToRemove);
```

Comment:
It would be! Though it occurs to me that the calling code and this code within the function don't maintain any consistency between who is supposed to check for the existence of the item to remove, with some calling locations checking for existence before calling, and some not checking

So it makes sense to me to get rid of the external existence checks and contain that to a single check here so callers can safely call this function, and do so as you say, with both remove calls behind the if check 👍